src/System/PairExchange.jl

```diff
@@ -92,11 +92,16 @@ function proportionality_factor(A, B; atol=1e-12)
     end
 end
 
-function decompose_general_coupling(op, N1, N2; extract_parts)
+function decompose_general_coupling(op, N1, N2; extract_parts, gen=nothing)
     @assert size(op) == (N1*N2, N1*N2)
 
-    gen1 = spin_matrices_of_dim(; N=N1)
-    gen2 = spin_matrices_of_dim(; N=N2)
+    if isnothing(gen)
+        gen1 = spin_matrices_of_dim(; N=N1)
+        gen2 = spin_matrices_of_dim(; N=N2)
+    else
+        extract_parts = false
+        gen1 = gen2 = gen
```

Comment:
In principle `gen1` and `gen2` can be different (i.e., at least this function allows `N1` \neq `N2`). Should this be supported at the level of passing custom generators? (E.g., I imagine what you have will support hacking in the two doublet situation you were describing, but presumably one might encounter other situations.)

Would it be a good idea to to add an assertion check along the lines `size(gen) .^ 2 == size(op)`?

Mostly asking to make sure I'm following. I'm not really encouraging further complication.

Reply:
Yes, that check looks right. It probably happens elsewhere, e.g. when unitaries are applied.

This PR could alternatively lift both `gen1` and `gen2` to the public API. But since Sunny currently requires the same Hilbert space dimension _N_ on every site, this would suggest a flexibility that doesn't actually fully exist.

The presence of `gen1` and `gen2` in the current codebase is a historical accident. They were probably introduced as part of the scaffolding to transition to a better implementation, where the generators of rotations would be living in the `System` structure. That effort got postponed due to growing complexity.